### PR TITLE
SOT-154: Implement aggregate_commitments and batch_verify for KZG

### PR DESCRIPTION
**Pull request by hduoc2003**
- Aggregates multiple commitments into one commitment using a random challenge
- Verifies that each proof is a valid proof of evaluation for `commitment_i` at `point_i`

### CODE REVIEW

File: kzg/src/commitment.rs [hunk 1]
```diff
@@ -15,10 +15,14 @@ impl KzgCommitment {
 mod tests {
     use std::ops::Mul;
 
+    use crate::commitment::KzgCommitment;
```
Comment (zk-steve) [inline kzg/src/commitment.rs]: re-arrange the imports

File: kzg/src/commitment.rs [hunk 2]
```diff
@@ -57,12 +61,59 @@ mod tests {
...
+            Fr::from(2003),
+            Fr::from(13),
+            Fr::from(9),
+        ];
+        let openings: Vec<KzgOpening> = f
+            .iter()
+            .zip(z)
+            .map(|(f_i, z_i)| scheme.open(f_i.clone(), z_i))
+            .collect();
+        let c: Vec<KzgCommitment> = f.iter().map(|f_i| scheme.commit(f_i)).collect();
```
Comment (zk-steve) [inline kzg/src/commitment.rs:116]:
```suggestion
        let c: Vec<KzgCommitment> = f.iter().map(scheme.commit).collect();
```

File: kzg/src/scheme.rs [hunk 3]
```diff
@@ -124,6 +128,80 @@ impl KzgScheme {
...
+            let cm_minus_s = *cm.inner() - g1.mul(s).into_affine();
+            // z_i * w_i
+            let z_mul_w = *w * z;
+            // r'
+            let r_prime = Fr::from(rng.gen::<u128>());
+            // e_1 += r_prime^i * (cm_i - s_i + z_i * w_i)
+            e_1 = Affine::from(e_1 + (cm_minus_s + z_mul_w) * r_prime);
+
+            // e_2 += r_prime^i * w_i;
+            e_2 = Affine::from(e_2 + (*w * r_prime));
```
Comment (ktokto313) [inline kzg/src/scheme.rs:198]: why e_1 and e_2 need to be initiated as 0 and then += ?

File: kzg/src/scheme.rs [hunk 4]
```diff
@@ -124,6 +128,80 @@ impl KzgScheme {
...
+    ) -> bool {
+        assert_eq!(commitments.len(), points.len());
+        assert_eq!(openings.len(), points.len());
+
+        let g1 = G1Point::generator();
+        let mut rng = StdRng::from_entropy();
+        let mut e_1 = G1Point::zero();
+        let mut e_2 = G1Point::zero();
+
+        for ((cm, z), KzgOpening(w, s)) in commitments.iter().zip(points).zip(openings) {
```
Comment (ktokto313) [inline kzg/src/scheme.rs:187]: can you write doc for this part, it's really hard for my brain to interpret what cm, z, w and s is every step

File: kzg/src/scheme.rs [hunk 5]
```diff
@@ -124,6 +128,80 @@ impl KzgScheme {
...
+            // cm_i - s_i
+            let cm_minus_s = *cm.inner() - g1.mul(s).into_affine();
+            // z_i * w_i
+            let z_mul_w = *w * z;
+            // r'
+            let r_prime = Fr::from(rng.gen::<u128>());
```
Comment (ktokto313) [inline kzg/src/scheme.rs:193]: this part need more self-explanatory comment 
like cm_i - s_i should be g1 * f(s) - f(z)

**Review by ktokto313 — COMMENTED** [review]
Not similar to paper

File: kzg/src/commitment.rs (hunk 2 above)
Comment (zk-steve) [reply kzg/src/commitment.rs:116]: commit first before opening

**Review by ktokto313 — APPROVED** [review]

**Pull request merged**